### PR TITLE
Handle case where the hole vertex is south of the containing polygon(s)

### LINKED ISSUE #25933: Geo_shape indexing issue - java.lang.ArrayIndexOutOfBoundsException
**Elasticsearch version**: 5.4.1

**Plugins installed**: []

**JVM version** (`java -version`): java version "1.8.0_131"

**OS version** (`uname -a` if on a Unix-like system): MacOS 10.11.6

**Description of the problem including expected versus actual behavior**:
Have been trying to index a lot of geometry, and getting consistently the same problems with those geometry/polygons that seem to contain a lot of vertices. Have tried importing the shapefile, and geojson through ogr2ogr. also tried generating a json file to use with bulk update. The error I am getting is an ArrayOutofBoundsException. 

Also to note, if I upload the JSON file without specifying a mapping i.e. a geo_shape, then the data gets sucked in, but I would expect it's then not indexed.

As I have consistently been able to reproduce this with different upload routes, I am assuming it has to be a bug - is there a way around it? 

Error I am getting is below:
https://gist.github.com/aj7/12fa99044901d7db8b42a5d5d5753bf8

I am happy to attach the json file if someone want's to try and see for themselves?

### PR DESCRIPTION
**Pull request by DaveCTurner**
Normally the hole is assigned to the component of the first edge to the south
of one of its vertices, but if the chosen hole vertex is south of everything
then the binary search returns -1 yielding an `ArrayIndexOutOfBoundsException`.
Instead, assign the vertex to the component of the first edge to its north.
Subsequent validation catches the fact that the hole is outside its component.

Fixes #25933

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java [hunk 1]
```diff
@@ -124,4 +125,24 @@ public void testCoerceHole() {
         assertThat("hole should have been closed via coerce", pb.holes().get(0).coordinates(false).length, equalTo(4));
     }
 
+    public void testHoleThatIsSouthOfPolygon() {
+        try {
+            PolygonBuilder pb = new PolygonBuilder(new CoordinatesBuilder().coordinate(4, 3).coordinate(3, 2).coordinate(3, 3).close());
+            pb.hole(new LineStringBuilder(new CoordinatesBuilder().coordinate(4, 2).coordinate(3, 1).coordinate(4, 1).close()));
+            pb.build();
+        } catch (InvalidShapeException e) {
+            assertEquals("Hole lies outside shell at or near point (4.0, 1.0, NaN)", e.getMessage());
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java]: nit: there is the neat expectThrows() utility in LuceneTestCase that makes test for exceptions a little more succinct. MIght help here (and in the other test) as well

**Review by cbuescher — COMMENTED** [review]
@DaveCTurner I had a look at the PR and left a small comment, other than that I'm not too familiar with the validation of polygons wt. holes and cannot comment with much certainty on the proposed fix. I think @nknize should have a look at this as well.

File: core/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java (hunk 1 above)
Comment (DaveCTurner) [reply core/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java]: D'oh, I forgot about that. I pushed dc88dace9b

**Comment by DaveCTurner** [comment]
@nknize does this look ok to you?

File: core/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java [hunk 2]
```diff
@@ -482,7 +482,7 @@ private static void assign(Edge[] holes, Coordinate[][] points, int numHoles, Ed
                             && !(sharedVertex = (edges[pos].intersect.compareTo(current.coordinate) == 0)) ) {
                 throw new InvalidShapeException("Invalid shape: Hole is not within polygon");
             }
-            final int index = -((sharedVertex) ? 0 : pos+2);
+            final int index = sharedVertex || pos == -1 ? 0 : -(pos+2);
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java]: @DaveCTurner thanks for walking me trough this f2f, it took some time for me to understand but now I think I understand this whole method a lot better. This looks right to me now after some longer thought, I think we should add some comments to this part after that explains the three(?) cases that might be true at this point and what that means for the index of the edge we assign this component to.

**Review by cbuescher — CHANGES_REQUESTED** [review]

**Review by cbuescher — APPROVED** [review]
@DaveCTurner thanks a lot for the additional comments. More verbose but definitely more readable IMHO.

**Comment by DaveCTurner** [comment]
Merged as e6da564 and backported to 6.x as ba83edac0c.

**Pull request merged**